File: packs/dsa-trees-ts/challenges/08-merge-two-binary-trees.test.ts
import { join } from 'node:path';
import { writeChallengeToTmp, importModule } from '../../test-helpers.js';

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function buildTree(values: (number | null)[]): TreeNode | null {
  if (!values.length || values[0] === null) return null;
  const root = new TreeNode(values[0]);
  const queue: TreeNode[] = [root];
  let i = 1;
  while (queue.length && i < values.length) {
    const node = queue.shift()!;
    if (i < values.length && values[i] !== null) {
      node.left = new TreeNode(values[i] as number);
      queue.push(node.left);
    }
    i++;
    if (i < values.length && values[i] !== null) {
      node.right = new TreeNode(values[i] as number);
      queue.push(node.right);
    }
    i++;
  }
  return root;
}

function treeToList(root: TreeNode | null): (number | null)[] {
  if (!root) return [];
  const result: (number | null)[] = [];
  const queue: (TreeNode | null)[] = [root];
  while (queue.length) {
    const node = queue.shift()!;
    if (node) {
      result.push(node.val);
      queue.push(node.left);
      queue.push(node.right);
    } else {
      result.push(null);
    }
  }
  while (result.length && result[result.length - 1] === null) result.pop();
  return result;
}

const CHALLENGE = join(import.meta.dirname, '08-merge-two-binary-trees.json');

describe('Merge Two Binary Trees', () => {
  let mergeTrees: (
    root1: TreeNode | null,
    root2: TreeNode | null
  ) => TreeNode | null;
  let cleanup: () => void;

  beforeAll(async () => {
    const tmp = writeChallengeToTmp(CHALLENGE);
    cleanup = tmp.cleanup;
    const mod = await importModule<{ mergeTrees: typeof mergeTrees }>(
      join(tmp.tmpDir, 'solution.ts')
    );
    mergeTrees = mod.mergeTrees;
  });

  afterAll(() => cleanup());

  it('merges overlapping trees', () => {
    const result = mergeTrees(
      buildTree([1, 3, 2, 5]),
      buildTree([2, 1, 3, null, 4, null, 7])
    );
    expect(treeToList(result)).toEqual([3, 4, 5, 5, 4, null, 7]);
  });

  it('merges trees with different shapes', () => {
    const result = mergeTrees(buildTree([1]), buildTree([1, 2]));
    expect(treeToList(result)).toEqual([2, 2]);
  });

  it('returns second tree when first is null', () => {
    const result = mergeTrees(null, buildTree([1, 2, 3]));
    expect(treeToList(result)).toEqual([1, 2, 3]);
  });

  it('returns first tree when second is null', () => {
    const result = mergeTrees(buildTree([1, 2, 3]), null);
    expect(treeToList(result)).toEqual([1, 2, 3]);
  });

  it('returns null when both are null', () => {
    expect(mergeTrees(null, null)).toBeNull();
  });

  it('merges single nodes', () => {
    const result = mergeTrees(buildTree([5]), buildTree([3]));
    expect(treeToList(result)).toEqual([8]);
  });
});

File: packs/dsa-trees-py/challenges/08-merge-two-binary-trees.test.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree(values):
    """Build binary tree from level-order list. None means no node."""
    if not values or values[0] is None:
        return None
    root = TreeNode(values[0])
    queue = deque([root])
    i = 1
    while queue and i < len(values):
        node = queue.popleft()
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    return root


def tree_to_list(root):
    """Convert tree to level-order list (with None for missing nodes)."""
    if not root:
        return []
    result = []
    queue = deque([root])
    while queue:
        node = queue.popleft()
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    while result and result[-1] is None:
        result.pop()
    return result


from solution import merge_trees


def test_merge_overlapping():
    root1 = build_tree([1, 3, 2, 5])
    root2 = build_tree([2, 1, 3, None, 4, None, 7])
    result = merge_trees(root1, root2)
    assert tree_to_list(result) == [3, 4, 5, 5, 4, None, 7]


def test_merge_different_shapes():
    root1 = build_tree([1])
    root2 = build_tree([1, 2])
    result = merge_trees(root1, root2)
    assert tree_to_list(result) == [2, 2]


def test_merge_with_none_first():
    root2 = build_tree([1, 2, 3])
    result = merge_trees(None, root2)
    assert tree_to_list(result) == [1, 2, 3]


def test_merge_with_none_second():
    root1 = build_tree([1, 2, 3])
    result = merge_trees(root1, None)
    assert tree_to_list(result) == [1, 2, 3]


def test_merge_both_none():
    result = merge_trees(None, None)
    assert result is None


def test_merge_single_nodes():
    root1 = build_tree([5])
    root2 = build_tree([3])
    result = merge_trees(root1, root2)
    assert tree_to_list(result) == [8]
